### PR TITLE
feat: propagate model token limits

### PR DESCRIPTION
**Pull request by Alishahryar1**
## Problem

Provider model catalogs exposed context and output limits that FCC discarded, so downstream harnesses used generic defaults even when exact per-model values were available.

## Changes

| Before | After |
| --- | --- |
| Provider token-limit fields were discarded during model discovery. | Documented positive token limits flow through the canonical model metadata record. |
| FCC model-list responses omitted context and output limits. | Messages and Responses catalogs expose known limits and omit unknown values. |
| Harnesses relied on universal or incomplete token defaults. | Codex, Pi, OpenCode, Cline, Hermes, DeepSeek Harness, and Aider receive matching native limit fields. |
| Malformed optional limits could not be represented independently. | Strict parsing degrades each malformed or missing limit to unknown without dropping the model or its other metadata. |

<details open><summary><h3>Greptile Summary</h3></summary>

This update clarifies OpenCode’s paired token-limit representation. Models with only one known limit continue to use OpenCode’s zero sentinel for the unknown value, allowing OpenCode to apply its configured output-token fallback rather than imposing a zero-token generation limit.
</details>

<h3>Confidence Score: 5/5</h3>

No blocking failure remains: context-only and output-only model configurations produce OpenCode’s supported paired-limit representation, and the zero output sentinel resolves to the configured fallback.

No accepted blocking findings remain.

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- Ran the token-sentinel regression script with the before revision to verify that the prior implementation emitted unpaired, one-sided limits.
- Ran the token-sentinel regression script with the after revision to verify that the current implementation emits zero sentinels and caps the effective output at 32000.
- Ran the pytest suite for the opencode launcher and confirmed that all 37 tests passed, showing zero is interpreted as OpenCode’s fallback sentinel rather than a zero-token generation cap.

<a href="https://app.greptile.com/trex/runs/21366151/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<sub>Reviews (4): Last reviewed commit: ["docs: clarify OpenCode limit sentinel"](https://github.com/alishahryar1/free-claude-code/commit/eb381931781794d301130b9053af165376b14c4c) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=58159693)</sub>

### CODE REVIEW

File: src/free_claude_code/cli/launchers/opencode_config.py [hunk 1]
```diff
@@ -73,4 +73,9 @@ def _model_config(model: ClientModel) -> JsonObject:
                 if modality in model.input_modalities
             ]
         }
+    if model.context_window_tokens is not None or model.max_output_tokens is not None:
+        config["limit"] = {
+            "context": model.context_window_tokens or 0,
+            "output": model.max_output_tokens or 0,
+        }
```
[the remark this answers is not in the document]
Comment (Alishahryar1) [reply src/free_claude_code/cli/launchers/opencode_config.py:87]: The `0` values are intentional and match OpenCode v1.18.18's native unknown-limit representation. Its provider loader defaults missing context/output limits to `0` (`model.limit?.output ?? existingModel?.limit?.output ?? 0`), and `maxOutputTokens()` explicitly converts a zero output limit to `outputTokenMax` via `Math.min(model.limit.output, outputTokenMax) || outputTokenMax`. The config schema also requires both `context` and `output` when `limit` is present, so omitting the unknown sibling creates an invalid partial object. I kept the complete pair, made the sentinel intent explicit in code, and retain coverage for context-only and output-only models.

Sources: https://github.com/anomalyco/opencode/blob/v1.18.18/packages/opencode/src/provider/provider.ts#L1218-L1233 and https://github.com/anomalyco/opencode/blob/v1.18.18/packages/opencode/src/provider/transform.ts#L1420-L1422

**Comment by Alishahryar1** [comment]
@greptile review

File: src/free_claude_code/cli/launchers/opencode_config.py [hunk 2]
```diff
@@ -73,4 +73,16 @@ def _model_config(model: ClientModel) -> JsonObject:
...
+        config["limit"] = {
+            "context": (
+                model.context_window_tokens
+                if model.context_window_tokens is not None
+                else 0
+            ),
+            "output": (
+                model.max_output_tokens if model.max_output_tokens is not None else 0
+            ),
+        }
```
[the remark this answers is not in the document]
Comment (Alishahryar1) [reply src/free_claude_code/cli/launchers/opencode_config.py:87]: This duplicates the earlier finding that Greptile already withdrew after checking OpenCode v1.18.18. OpenCode's own provider loader uses `0` for missing limits, its regression test requires that default, and `maxOutputTokens()` converts a zero output limit to `outputTokenMax`; it does not send a zero-token allowance upstream. The complete pair is also required by the v1 config schema. The current projection is therefore the OpenCode-supported unknown representation, and no code change is appropriate.

Sources: https://github.com/anomalyco/opencode/blob/v1.18.18/packages/opencode/src/provider/provider.ts#L1218-L1233, https://github.com/anomalyco/opencode/blob/v1.18.18/packages/opencode/test/provider/provider.test.ts#L1232-L1244, and https://github.com/anomalyco/opencode/blob/v1.18.18/packages/opencode/src/provider/transform.ts#L1420-L1422
